R/add_missing_events.R

```diff
@@ -0,0 +1,221 @@
+# Author: Paigan Aspinall
+# Date: 09JAN2024
+# R version: 4.2.2
+
+#' Identify events missing from the dataset and, if their expected event date has passed, add them to the data.
+#'
+#' This function adds a row for any event that is expected in the data and is missing. NOTE this script should be run after allocation_data_cleaning and before encode_site_names.
+#'
+#' @param dataset A REDCap export dataset.
+#' @param event_data A CSV file containing the events and their associated anchors - template located in Github named "20240109_AddMissingEventsTemplate.csv".
+#' @param condition Optional parameter that can be applied, e.g. if the events for intervention and control participants are different. Default is NULL.
+#' @param allocation Logical, if allocation sare used in the study then this should be true. Default to TRUE.
+#' @param site Logical, if sites are used in the study then this should be true. Default to TRUE.
+
+#' @return A data frame containing missing events.
+#'
+#'
+#' @importFrom dplyr "%>%"
+#' 
+#' @examples:
+#' Example usage:
+#' intervention_events <- read.csv("20240108_ExpectedEventsInvervention_V1.0.csv")
+#' condition <- "Allocation == 'Intervention'"
+#' 
+#' complete_dataset <- add_missing_events(dataset, event_data, condition)
+#'
+#' @export
+#'
+
+add_missing_events <- function(dataset, event_data, condition=NULL, allocation=TRUE, site=TRUE){
+  
+  #calculate the upper limit of days after the anchor date
+  event_data$days_total <- event_data$days_after + event_data$days_upper
+  
+  #if a condition is being used, filter based on this
+  if(!is.na(condition)){
+  filtered_data <- dataset %>% filter(eval(parse(text = condition)))}
+  filtered_data$redcap_event_name <- as.character(filtered_data$redcap_event_name)
+
+  #group the data based on record id
+  grouped_large <- filtered_data %>%
+    group_by(record_id)
+  
+  #create a data frame to store all combinations of record_id and event_name
+  event_names <- unique(event_data$event_name)
+  combinations <- grouped_large %>%
+    summarize(record_id = first(record_id)) %>%
+    crossing(redcap_event_name = event_names)
+  
+  #merge necessary columns from event_data with combinations based on redcap_event_name
+  combinations <- merge(combinations, event_data, by.x = "redcap_event_name", by.y="event_name", all.x=TRUE )
+  names(filtered_data)[which(names(filtered_data)=="redcap_event_name")] <- "anchor_event"
+  combinations <- merge(combinations, filtered_data, by=c("record_id", "anchor_event"), all.x=TRUE)
+
+  #function to replace the 'anchor_date' column with the corresponding date value
+  replace_with_date <- function(row) {
```

Comment:
Don't love this being nested within another function. May be worth putting in a separate file but don't 'export' it as part of the package and just import into this function at the top.

Reply:
Updated.